docs/architecture/adr-061-core-module-api.md

```diff
@@ -0,0 +1,562 @@
+we haven't anticipated.
+
+#### Begin and End Blockers
+
+The `BeginBlock` and `EndBlock` methods will simply take a `context.Context`, because:
+* most modules don't need Tendermint information other than `BlockInfo` so we can eliminate dependencies on specific
+Tendermint versions
+* for the few modules that need Tendermint block headers and/or return validator updates, specific versions of the
+runtime module will provide specific functionality for interacting with the specific version(s) of Tendermint
+supported
+
+In order for `BeginBlock`, `EndBlock` and `InitGenesis` to send back validator updates and retrieve full Tendermint
+block headers, the runtime module for a specific version of Tendermint could provide services like this:
+```go
+type ValidatorUpdateService interface {
+	SetValidatorUpdates(context.Context, []abci.ValidatorUpdate)
+}
+
+type BeginBlockService interface {
+	GetBeginBlockRequest(context.Context) abci.RequestBeginBlock 
+}
+```
+
+We know these types will change at the Tendermint level and that also a very limited set of modules actually need this
+functionality, so they are intentionally kept out of core to keep core limited to the necessary, minimal set of stable
+APIs.
+
+#### Event Listeners
+
+Handlers allow event listeners for typed events to be registered that will be called deterministically whenever those
+events are called in the state machine. Event listeners are functions that take a context, the protobuf message type of
+the event and optionally return an error. If an event listener returns a non-nil error, this error will fail the process
+that emitted the event and revert any state transitions that this process would have caused.
+
+Event listeners for typed events can be added to the handler struct either by populating the `EventListeners` field
+or using generic helper methods:
+```go
+appmodule.AddEventListener(handler, func (ctx context.Context, e *EventReceive)) { ... })
+```
+
+Event listeners provide a standardized alternative to module hooks, such as `StakingHooks`, even though these hooks
+are still supported and described in [ADR 057: App Wiring](./adr-057-app-wiring-1.md)
+
+Only one event listener per event type can be defined per module. Event listeners will be called in a deterministic
+order based on alphabetically sorting module names. If a more customizable order is desired, additional parameters
+to the runtime module config object can be added to support optional custom ordering.
+
+Only events emitted using `EventManager.Emit` are observable using event listeners.
+
+#### Upgrade Handlers
+
+Upgrade handlers can be specified using the `UpgradeHandlers` field that takes an array of `UpgradeHandler` structs:
+```go
+type UpgradeHandler struct {
+	FromModule protoreflect.FullName
+	Handler func(context.Context) error
+}
+```
+
+In the [ADR 057: App Wiring](./adr-057-app-wiring-1.md) it is specified that each version of each module should be
```

Comment:
can you provide an example of non DI way of doing this.

Reply:
Will get to that... It will be easier once codegen is wrapped up because there are some API changes needed to our current setup